Add constant-folding for unary NVVM intrinsics

LewisCrawford opened the pull request:
Add support for constant-folding numerous NVVM unary arithmetic intrinsics (including f, d, and ftz_f variants):
  - nvvm.ceil.*
  - ~~nvvm.cos.approx.*~~
  - ~~nvvm.ex2.approx.*~~
  - nvvm.fabs.*
  - nvvm.floor.*
 -  ~~nvvm.lg2.approx.*~~
  - nvvm.rcp.*
  - nvvm.round.*
 -  ~~nvvm.rsqrt.approx.*~~
  - nvvm.saturate.*
 -  ~~nvvm.sin.approx.*~~
  - nvvm.sqrt.f
  - nvvm.sqrt.rn.*
  - ~~nvvm.sqrt.approx.*~~

LewisCrawford committed:
Add constant-folding for unary NVVM intrinsics

Add support for constant-folding numerous NVVM unary arithmetic
intrinsics (including f, d, and ftz_f variants):
  - nvvm.ceil.*
  - nvvm.cos.approx.*
  - nvvm.ex2.approx.*
  - nvvm.fabs.*
  - nvvm.floor.*
  - nvvm.lg2.approx.*
  - nvvm.rcp.*
  - nvvm.round.*
  - nvvm.rsqrt.approx.*
  - nvvm.saturate.*
  - nvvm.sin.approx.*
  - nvvm.sqrt.f
  - nvvm.sqrt.rn.*
  - nvvm.sqrt.approx.*

LewisCrawford commented:
I would like to land this patch: https://github.com/llvm/llvm-project/pull/140270 before merging this. That patch adds an option to disable constant-folding for FP function calls, and may be useful in the few cases where the device-side execution of these intrinsics must be a bitwise-exact match to the host-side constant-folded version of the results (or when debugging).

LewisCrawford committed:
Fix _approx_ftz_d implementations for rcp + rsqrt

Fix the rcp_approx_ftz_d and rsqrt_approx_ftz_d implementations
to better match the PTX spec, which states that the inputs and outputs
should zero the lower 32-bits of the mantissa.

Artem-B commented on llvm/lib/Analysis/ConstantFolding.cpp:
```diff
@@ -1944,16 +2006,32 @@ static const APFloat FTZPreserveSign(const APFloat &V) {
+}
+
+Constant *ConstantFoldFP(double (*NativeFP)(double), const APFloat &V, Type *Ty,
+                         bool ShouldFTZPreservingSign = false) {
```
I'm curious why sign-preserving FTZ is the special case? Is a non-sign-preserving FTZ a thing?

Afaict, all we care here is should we FTZ or not, and sign-preserving FTZ is the only one we have implemented in const-folding code. Perhaps the parameter should be called just `doFTZ` ?

Artem-B replied:
Interesting. I wasn't aware of that. Do we need to make the parameter an enum (none, ftz, ftz-positive) then? It sounds like constant folding should change its behavior based on the `-fdenormal-fp-math`, too.

LewisCrawford committed:
Merge branch 'main' into fold_nvvm_math_intrinsics

LewisCrawford committed:
Fix rcp.rp rounding modes

RCP incorrectly folded rcp.rp as rcp.rn due to a bug
in NVVMIntrinsicUtils.h.

LewisCrawford committed:
Remove returns after llvm_unreachable

Remove "return" statements providing default values
after llvm_unreachable calls in NVVMIntrinsicUtils.h.

LewisCrawford committed:
Handle arbitrary denormal in/out modes

Change ConstantFoldFP to accept an optional DenormalMode
parameter, which specifies how denormal inputs and outputs
are handled.

Avoid handling for invalid/dynamic modes, and allow for
flushing to either positive or sign-preserving zero (which
may be different for input vs output modes).

Allow the NVVM intrinsics to be folded with either the
IEEE mode (preserves subnormals), or the sign-preserving-zero
mode depending on whether the FTZ variant of the intrinsic is used.

LewisCrawford committed:
Fix clang-format error

LewisCrawford committed:
Merge branch 'main' into fold_nvvm_math_intrinsics

LewisCrawford committed:
Guard nvvm intrinsic folding with StrictFP check

LewisCrawford commented on llvm/lib/Analysis/ConstantFolding.cpp:
```diff
@@ -2548,6 +2653,170 @@ static Constant *ConstantFoldScalarCall1(StringRef Name,
+            nvvm::GetNVVMDenromMode(
+                nvvm::UnaryMathIntrinsicShouldFTZ(IntrinsicID)));
+
+      case Intrinsic::nvvm_cos_approx_ftz_f:
+      case Intrinsic::nvvm_cos_approx_f:
+        return ConstantFoldFP(
+            cos, APF, Ty,
+            nvvm::GetNVVMDenromMode(
+                nvvm::UnaryMathIntrinsicShouldFTZ(IntrinsicID)));
+
+      case Intrinsic::nvvm_ex2_approx_ftz_f:
+      case Intrinsic::nvvm_ex2_approx_d:
+      case Intrinsic::nvvm_ex2_approx_f:
+        return ConstantFoldFP(
+            exp2, APF, Ty,
+            nvvm::GetNVVMDenromMode(
+                (nvvm::UnaryMathIntrinsicShouldFTZ(IntrinsicID))));
```
I believe it is ok to do so, and I added the flag `-disable-fp-call-folding` earlier in https://github.com/llvm/llvm-project/pull/140270 sepcifically to allow users who need the bit-exact evaluations of these functions to match. Also, StrictFP provides another way to prevent these from being folded if necessary.

I believe in the vast majority of cases, users won't be requiring bit-exact evaluation of `approx` intrinsics, in the same way that users should not rely on specific fma fusion behaviour if -fast-math is enabled, and should use the non-approx versions of these intrinsics if they require exact results.

The LLVM versions of these intrinsics like llvm.sin, llvm.exp2 etc. are already folded using math library calls in this way, and can also cause similar precision mismatches e.g. in cross-compilation scenarios where the device the code was compiled on has a different architecture from the target the code executes on (or even a different libm implementation).

Artem-B replied:
> I believe in the vast majority of cases, users won't be requiring bit-exact evaluation of approx intrinsics,

The common failure pattern is that a user may have a test checking for the "golden" values, and that test will be passing/failing depending on the optimization level. They may not even be aware that an imprecise intrinsic was used somewhere under the hood. Granted, those tests usually come with precision/tolerance knobs, and can usually be fixed by relaxing the expected range of results, but I do expect to see some amount of unexpected trouble on the user's end. 

Probably not a showstopper.

> The LLVM versions of these intrinsics like llvm.sin, llvm.exp2 etc. are already folded using math library calls in this way, and can also cause similar precision mismatches

I think the difference here is in the degree of the mismatches. Both LLVM intrinsics and the math library are expected to give you very close results. Maybe not identical, but close (@jhuber6 are there any formal guarantees on that?). However, many GPU intrinsics have higher ULP errors and are not IEEE compliant.
https://docs.nvidia.com/cuda/cuda-c-programming-guide/#mathematical-functions-appendix

Is there a specific limit on how much of a difference is OK? I assume that the "too much of a difference" boundary does exist and that we're not crossing it.

LewisCrawford replied:
You're right about -O0 not folding. There are no checks for optimization level in ConstantFolding.cpp, and IRBuilder folds many instructions by default when they are created, but it looks like -O0 manages to avoid most const-folding by not calling InstCombine and by the initial parser creating the instructions directly without going through IRBuilder.

I agree that technically the semantics of the front-end shouldn't matter here, as any frontend can theoretically generate these. However, I cannot find any documentation for `nvvm.sin.approx` at the IR level, so am left to infer its semantics from the Cuda spec, NVCC behaviour that generates it, and the PTX spec for `sin.approx` which states it "implements a fast approximation to sine", and only gives max error bounds. From all of these sources, I'd argue that the "approx" in "nvvm.sin.approx" is essentially same as having the "afn" fast-math flag encoded in to the intrinsic's name. This would always imply that users of the intrinsic cannot rely on its precision, and would prefer speed over accuracy. Front-ends should only generate these fast "`approx`" intrinsics in cases where a fast imprecise approximation is expected.

Would the difference between unfolded O0 and folded O3 results be ok if there was an explicit `afn` fast-math flag attached to these intrinsics? Is having the "approx" included in the instrinsic name drastically different than using the separate `afn` flag?

> It may be OK, but I do not know where is the boundary on what's tolerable in this case. I'm biased towards keeping the default behavior unchanged and enable imprecise folding only if we have an unambiguous signal that it's OK to do so.

I'm also not sure what the boundary for what would be tolerable in this case would be. I'd argue that the "`approx`" in the name of the intrinsics is enough of a signal, but I'm probably biased in the direction of folding as much as possible. If anyone has more expertise about what an acceptable error would be for this to deviate between device-side execution with O0 and host-side folding with O3, then I'd be fine to add some limitations e.g. only fold on a restricted range within error tolerances if necessary.

arsenm replied:
Mostly I view the semantics of intrinsics as "gives the result the machine will give", not "perform mathematical operation"

arsenm replied:
We also shouldn't fold sin and cos that way

LewisCrawford replied:
Ok, I have removed all "approx" intrinsic folding from the patch now.

LewisCrawford committed:
Remove all approximate intrinsic folding

Artem-B reviewed (approved):
LGTM with a test suggestion.
